feats/redis.py

```diff
@@ -0,0 +1,152 @@
+    the last value at the end of the loop
+    """
+    def __init__(self, stream):
+        self.index = 0
+        self.end = len(stream)
+        self.stream = stream
+
+    def __iter__(self):
+        self.index = 0
+        return self
+
+    def __next__(self):
+        if self.index >= self.end:
+            raise StopIteration
+        result = self.stream[self.index]
+        self.index += 1
+        return result[1]
+
+
+class FeatureStream:
+    """
+    This class provides utility methods for retrieving data from Redis.
+    It should not be initialized directly but instead returned from keying off
+    a RedisClient instance.
+    """
+    def __init__(self, redis, key):
+        self.key = key
+        self._redis = redis
+
+    def append(self, state) -> str:
+        """
+        Adds FeatureState data to the head of the Redis Stream and returns
+        the Redis auto-generated id
+        """
+        return self._redis.xadd(self.key, state)
+
+    def info(self):
+        """
+        Wrapper for redis xinfo
+        """
+        return self._redis.xinfo_stream(self.key)
+
+    def read(self, index) -> dict:
+        """
+        reads the stream for a given redis-generated key (eg '1576268467400-0')
+        and returns a single value, stripping off the id
+        """
+        return self.range(start=index, end=index)[0]
+
+    def range(self, start='-', end='+') -> List[dict]:
+        """
+        reads the stream for a range of ids, default is entire stream
+        """
+        return self._redis.xrange(self.key, min=start, max=end)
+
+    def last(self) -> dict:
+        """
+        Returns the value of the latest entry in the Redis Stream (omitting id)
+        """
+        info = self.info()
+        return info['last-entry'][1]
+
+    def first(self) -> dict:
+        """
+        Returns the value of the first entry in the Redis Stream (omitting id)
+        """
+        info = self.info()
+        return info['first-entry'][1]
+
+    def __len__(self) -> int:
+        """
+        Wrapper for stream xlen
+        """
+        return self._redis.xlen(self.key)
+
+    def __iter__(self) -> StreamIterator:
+        """
+        Fetches the entire Redis stream and loads it into memory then returns
+        a StreamIterator that will iterate over all values in the stream.
+        """
+        # Fetch the entire list and then iterate over it synchronously
+        stream = self.range()
+        return StreamIterator(stream)
+
+
+class RedisClient:
+    """
+    Wrapper class for the redis-py connection, provides some utility methods
+    to adhere to the Storage interface -- namely indexing on stream keys and
+    getting a Stream object with an append method back.
+
+    Initializing this class initializes the redis connection client but does
+    _not_ test the connection to redis server.
+    """
+    def __init__(self, host='localhost', port=6379, db=0, decode_responses=True, **options):
```

Comment:
When would one want decode_responses to be False?

Reply:
Not sure, but I just didn't want to make this argument unavailable to the user in case they wanted to for some reason. I'll update the initialization of the Redis client so they can override this value with `**options`